tests/fixtures/database.py

```diff
@@ -0,0 +1,1048 @@
+        # Destroy the database connection and engine.
+        self._connection.close()
+        self._engine.dispose()
+
+    def connection(self) -> Connection:
+        return self._connection
+
+
+class DatabaseTransactionFixture:
+    """A fixture representing a single transaction. The transaction is automatically rolled back."""
+
+    _database: DatabaseFixture
+    _default_library: Optional[Library]
+    _default_collection: Optional[Collection]
+    _session: Session
+    _transaction: Transaction
+    _counter: int
+    _isbns: List[str]
+
+    def __init__(
+        self, database: DatabaseFixture, session: Session, transaction: Transaction
+    ):
+        self._database = database
+        self._session = session
+        self._transaction = transaction
+        self._default_library = None
+        self._default_collection = None
+        self._counter = 2000
+        self._isbns = [
+            "9780674368279",
+            "0636920028468",
+            "9781936460236",
+            "9780316075978",
+        ]
+
+    def _make_default_library(self):
+        """Ensure that the default library exists in the given database."""
+        library, ignore = get_one_or_create(
+            self._session,
+            Library,
+            create_method_kwargs=dict(
+                uuid=str(uuid.uuid4()),
+                name="default",
+            ),
+            short_name="default",
+        )
+        collection, ignore = get_one_or_create(
+            self._session, Collection, name="Default Collection"
+        )
+        integration = collection.create_external_integration(
+            ExternalIntegration.OPDS_IMPORT
+        )
+        integration.goal = ExternalIntegration.LICENSE_GOAL
+        if collection not in library.collections:
+            library.collections.append(collection)
+        return library
+
+    @staticmethod
+    def create(database: DatabaseFixture) -> "DatabaseTransactionFixture":
+        # Create a new connection to the database.
+        session = Session(database.connection())
+        transaction = database.connection().begin_nested()
+        return DatabaseTransactionFixture(database, session, transaction)
+
+    def close(self):
+        # Close the session.
+        self._session.close()
+
+        # Roll back all database changes that happened during this
+        # test, whether in the session that was just closed or some
+        # other session.
+        self._transaction.rollback()
+
+        # Reset the Analytics singleton between tests.
+        Analytics._reset_singleton_instance()
+
+        # Also roll back any record of those changes in the
+        # Configuration instance.
+        for key in [
+            Configuration.SITE_CONFIGURATION_LAST_UPDATE,
+            Configuration.LAST_CHECKED_FOR_SITE_CONFIGURATION_UPDATE,
+        ]:
+            if key in Configuration.instance:
+                del Configuration.instance[key]
+
+    def database(self) -> DatabaseFixture:
```

Comment:
database, transaction, session, default_collection and default_library are used as attributes, why not define them as `@property`'s

Reply:
Don't `@property`s imply setters? These should be read-only.

Reply:
`@property` is a getter, whereas `@<property-name>.setter` is a setter
https://docs.python.org/3/library/functions.html#property

Reply:
I'm not sure I see the value of making them properties here. Could you elaborate a little?

I'm not opposed to doing it, it's just not clear to me what it gets us.

Reply:
There is no functional change between the `@property` or current method implementation, it is only syntactic sugar where the test cases would be accessing `.session` as opposed to `.session()`

Reply:
Refactored most of these to properties, and that allowed for inlining some local variables.